content/blog/2022/2022-06-30-configuring-cors-with-spring.md

```diff
@@ -0,0 +1,430 @@
+
+````java
+    @CrossOrigin(maxAge = 3600)
+    @RestController
+    @RequestMapping("cors-library/managed/books")
+    public class LibraryController {
+
+        private static final Logger log = LoggerFactory.getLogger(LibraryController.class);
+
+        private final LibraryService libraryService;
+
+        public LibraryController(LibraryService libraryService) {
+            this.libraryService = libraryService;
+        }
+
+        @CrossOrigin(origins = "http://localhost:4200", allowedHeaders = "Requestor-Type")
+        @GetMapping
+        public ResponseEntity<List<BookDto>> getBooks(@RequestParam String type) {
+            HttpHeaders headers = new HttpHeaders();
+            headers.set("X-Get-Header", "ExampleHeader");
+            return ResponseEntity.ok().headers(headers).body(libraryService.getAllBooks(type));
+        }
+    }
+````
+By defining the annotation at both class and method levels
+    - its combined attributes will be applied to the methods i.e (origins, allowedHeaders, maxAge)
+
+### Enable CORS Configuration globally
+
+Instead of adding CORS to each of the resource separately, we could define a common CORS configuration that would apply to
+all resources defined. Here, we will use **WebMvcConfigurer** which is a part of the Spring Web MVC library.
+By overriding the **addCorsMapping** we will configure CORS to all URLs that are handled by Spring Web MVC.
+
+To define the same configuration (as explained in the previous sections) globally, we will use the configuration parameters
+defined in **application.yml** to create a bean as defined below:
+
+````yaml
+    web:
+    cors:
+      allowed-origins: "http://localhost:4200"
+      allowed-methods: GET, POST, PATCH, PUT, DELETE, OPTIONS, HEAD
+      max-age: 3600
+      allowed-headers: "Requestor-Type"
+      exposed-headers: "X-Get-Header"
+````
+
+````java
+    @Bean
+    public WebMvcConfigurer corsMappingConfigurer() {
+            return new WebMvcConfigurer() {
+                @Override
+                public void addCorsMappings(CorsRegistry registry) {
+                    WebConfigProperties.Cors cors = webConfigProperties.getCors();
+                    registry.addMapping("/**")
+                    .allowedOrigins(cors.getAllowedOrigins())
+                    .allowedMethods(cors.getAllowedMethods())
+                    .maxAge(cors.getMaxAge())
+                    .allowedHeaders(cors.getAllowedHeaders())
+                    .exposedHeaders(cors.getExposedHeaders());
+            }
+        };
+    }
+````
+
+{{% info title="NOTE:" %}}
+**addMapping()** returns a **CorsRegistration** object which applies default **CorsConfiguration** if
+one or more methods(**allowedOrigins**, **allowedMethods**, **maxAge**, **allowedHeaders**, **exposedHeaders**) are not explicitly defined.
+Refer to the Spring library method **CorsConfiguration.applyPermitDefaultValues()** to understand the defaults applied.
```

Comment:
When you ask the reader to refer to some docs, it would be nice if you linked to them :)